File: weblayer/browser/content_filter_manager_impl.cc
// Copyright 2022 Neeva. All rights reserved.

#include "weblayer/browser/content_filter_manager_impl.h"

#include "base/callback.h"
#include "base/files/file_path.h"
#include "base/memory/ptr_util.h"
#include "components/neeva/browser/content_filter_rules_config.h"
#include "content/public/browser/browser_context.h"

#if BUILDFLAG(IS_ANDROID)
#include "base/android/callback_android.h"
#include "base/android/jni_string.h"
#include "base/android/scoped_java_ref.h"
#include "weblayer/browser/java/jni/ContentFilterManagerImpl_jni.h"
#endif

namespace weblayer {
namespace {

neeva::mojom::ContentFilterMode ToContentFilterMode(int mode) {
  switch (mode) {
    case 0:
      return neeva::mojom::ContentFilterMode::BLOCK_COOKIES;
    default:
      return neeva::mojom::ContentFilterMode::BLOCK_REQUESTS;
  }
}

}  // namespace

ContentFilterManagerImpl::ContentFilterManagerImpl(
    content::BrowserContext* browser_context)
    : browser_context_(browser_context) {}
ContentFilterManagerImpl::~ContentFilterManagerImpl() = default;

#if BUILDFLAG(IS_ANDROID)
void ContentFilterManagerImpl::SetRulesFile(
    JNIEnv* env, const base::android::JavaParamRef<jstring>& apk_path) {
  neeva::ContentFilterRulesConfig::GetOrCreate(browser_context_)->
      SetRulesFile(ConvertJavaStringToUTF8(apk_path));
}

void ContentFilterManagerImpl::SetMode(JNIEnv* env, int mode) {
  neeva::ContentFilterRulesConfig::GetOrCreate(browser_context_)->
      SetMode(ToContentFilterMode(mode));
}

void ContentFilterManagerImpl::AddHostExclusion(
    JNIEnv* env, const base::android::JavaParamRef<jstring>& host) {
  neeva::ContentFilterRulesConfig::GetOrCreate(browser_context_)->
      AddHostExclusion(ConvertJavaStringToUTF8(host));
}

void ContentFilterManagerImpl::RemoveHostExclusion(
    JNIEnv* env, const base::android::JavaParamRef<jstring>& host) {
  neeva::ContentFilterRulesConfig::GetOrCreate(browser_context_)->
      RemoveHostExclusion(ConvertJavaStringToUTF8(host));
}

void ContentFilterManagerImpl::ClearAllHostExclusions(JNIEnv* env) {
  neeva::ContentFilterRulesConfig::GetOrCreate(browser_context_)->ClearAllHostExclusions();
}

void ContentFilterManagerImpl::StartFiltering(JNIEnv* env) {
  neeva::ContentFilterRulesConfig::GetOrCreate(browser_context_)->StartFiltering();
}

void ContentFilterManagerImpl::StopFiltering(JNIEnv* env) {
  neeva::ContentFilterRulesConfig::GetOrCreate(browser_context_)->StopFiltering();
}
#endif

}  // namespace weblayer

File: weblayer/browser/java/org/chromium/weblayer_private/ContentFilterManagerImpl.java
// Copyright 2022 Neeva. All rights reserved.

package org.chromium.weblayer_private;

import android.webkit.ValueCallback;

import org.chromium.base.Callback;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.weblayer_private.interfaces.ContentFilterMode;
import org.chromium.weblayer_private.interfaces.IContentFilterManager;
import org.chromium.weblayer_private.interfaces.IObjectWrapper;
import org.chromium.weblayer_private.interfaces.ObjectWrapper;
import org.chromium.weblayer_private.interfaces.StrictModeWorkaround;

/**
 * Implementation of IContentFilterManager.
 */
@JNINamespace("weblayer")
public final class ContentFilterManagerImpl extends IContentFilterManager.Stub {
    private long mNativeContentFilterManager;

    ContentFilterManagerImpl(long nativeContentFilterManager) {
        mNativeContentFilterManager = nativeContentFilterManager;
    }

    public void destroy() {
        mNativeContentFilterManager = 0;
    }

    @Override
    public void enableRulesFile(String rules_file) {
        StrictModeWorkaround.apply();
        ContentFilterManagerImplJni.get().enableRulesFile(mNativeContentFilterManager, rules_file);
    }

    @Override
    public void disableAllRulesFiles() {
        StrictModeWorkaround.apply();
        ContentFilterManagerImplJni.get().disableAllRulesFiles(mNativeContentFilterManager);
    }

    @Override
    public void setMode(int mode) {
        StrictModeWorkaround.apply();
        ContentFilterManagerImplJni.get().setMode(mNativeContentFilterManager, mode);
    }

    @Override
    public void addHostExclusion(String host) {
        StrictModeWorkaround.apply();
        ContentFilterManagerImplJni.get().addHostExclusion(mNativeContentFilterManager, host);
    }

    @Override
    public void removeHostExclusion(String host) {
        StrictModeWorkaround.apply();
        ContentFilterManagerImplJni.get().removeHostExclusion(mNativeContentFilterManager, host);
    }

    @Override
    public void clearAllHostExclusions() {
        StrictModeWorkaround.apply();
        ContentFilterManagerImplJni.get().clearAllHostExclusions(mNativeContentFilterManager);
    }

    @Override
    public void startFiltering() {
        StrictModeWorkaround.apply();
        ContentFilterManagerImplJni.get().startFiltering(mNativeContentFilterManager);
    }

    @Override
    public void stopFiltering() {
        StrictModeWorkaround.apply();
        ContentFilterManagerImplJni.get().stopFiltering(mNativeContentFilterManager);
    }

    @NativeMethods
    interface Natives {
        void enableRulesFile(long nativeContentFilterManagerImpl, String rulesFile);
        void disableAllRulesFiles(long nativeContentFilterManagerImpl);
        void setMode(long nativeContentFilterManagerImpl, int mode);
        void addHostExclusion(long nativeContentFilterManagerImpl, String host);
        void removeHostExclusion(long nativeContentFilterManagerImpl, String host);
        void clearAllHostExclusions(long nativeContentFilterManagerImpl);
        void startFiltering(long nativeContentFilterManagerImpl);
        void stopFiltering(long nativeContentFilterManagerImpl);
    }
}
